src/api/apiErrors.js

```diff
@@ -0,0 +1,62 @@
+/* @flow strict-local */
+import type { ApiResponseErrorData } from './transportTypes';
+import * as logging from '../utils/logging';
+
+/**
+ * Runtime class of custom API error types.
+ */
+export class ApiError extends Error {
+  data: $ReadOnly<ApiResponseErrorData>;
+  httpStatus: number;
+
+  constructor(httpStatus: number, data: $ReadOnly<ApiResponseErrorData>) {
+    super(data.msg);
+    this.data = data;
+    this.httpStatus = httpStatus;
+  }
+}
+
+/**
+ * Given a server response (allegedly) denoting an error, produce an Error to be
+ * thrown.
+ *
+ * If the `data` argument is actually of the form expected from the server, the
+ * returned error will be an {@link ApiError}; otherwise it will be a generic
+ * Error.
+ */
+export const makeErrorFromApi = (httpStatus: number, data: ?mixed): Error => {
+  // Validate `data`, and construct the resultant error object.
+  if (data !== null && typeof data === 'object') {
+    if (typeof data.msg === 'string' && data.result === 'error') {
+      if ('code' in data && typeof data.code === 'string') {
+        /* $FlowFixMe, in the unlikely event we can ever narrow types with if
+        statements on properties. */
```

Comment:
nit: indent so it doesn't look like the start of a non-comment line

Perhaps just use `//` style comments? Actually, though, I'm (pleasantly) surprised that `$FlowFixMe` works when there's another comment line in between like this; not sure if that might vary with the comment style.

Reply:
> I'm afraid it does vary with the comment style. It seems to look for the next statement after the comment – including both compound statements and other comments.

Welp. Good to know that's the rule, though.